backends/gstreamer/src/audio_decoder.rs

```diff
@@ -97,84 +115,164 @@ impl AudioDecoder for GStreamerAudioDecoder {
                 return callbacks.error();
             }
 
-            let insert_sink = || -> Result<(), ()> {
-                let queue = gst::ElementFactory::make("queue", None).ok_or(())?;
+            let sample_audio_info = match gst_audio::AudioInfo::from_caps(&caps) {
+                Some(sample_audio_info) => sample_audio_info,
+                None => return callbacks.error(),
+            };
+            let channels = sample_audio_info.channels();
+            callbacks.ready(channels);
+
+            let insert_deinterleave = || -> Result<(), ()> {
                 let convert = gst::ElementFactory::make("audioconvert", None).ok_or(())?;
                 let resample = gst::ElementFactory::make("audioresample", None).ok_or(())?;
-                let sink = gst::ElementFactory::make("appsink", None).ok_or(())?;
-                let appsink = sink.clone().dynamic_cast::<AppSink>().map_err(|_| ())?;
-                sink.set_property("sync", &false.to_value())
+                let filter = gst::ElementFactory::make("capsfilter", None).ok_or(())?;
+                let deinterleave =
+                    gst::ElementFactory::make("deinterleave", Some("deinterleave")).ok_or(())?;
+
+                deinterleave
+                    .set_property("keep-positions", &true.to_value())
                     .map_err(|_| ())?;
+                let pipeline_ = pipeline.clone();
+                let callbacks_ = callbacks.clone();
+                deinterleave.connect_pad_added(move |_, src_pad| {
```

Comment:
You create a reference cycle here: pipeline -> deinterleave -> pipeline.

Best to use weak references (`pipeline.downgrade()` and later `pipeline.upgrade()`) for passing into the closure. Same bug elsewhere probably